tests/bsim/bluetooth/host/l2cap/credits_seg_recv/src/main.c

```diff
@@ -0,0 +1,450 @@
+/*
+ * Copyright (c) 2022 Nordic Semiconductor
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "bstests.h"
+#include "common.h"
+
+#define LOG_MODULE_NAME main
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(LOG_MODULE_NAME, LOG_LEVEL_DBG);
+
+CREATE_FLAG(is_connected);
+CREATE_FLAG(flag_l2cap_connected);
+
+#define L2CAP_MPS CONFIG_BT_L2CAP_TX_MTU
+#define SDU_NUM   3
+#define SDU_LEN   (2 * L2CAP_MPS)
+/* We intentionally send smaller SDUs than the channel can fit */
+#define L2CAP_MTU (2 * SDU_LEN)
+
+/* Only one SDU transmitted or received at a time */
+NET_BUF_POOL_DEFINE(sdu_pool, 1, L2CAP_MTU, 8, NULL);
+
+static uint8_t tx_data[SDU_LEN];
+static uint16_t rx_cnt;
+
+K_SEM_DEFINE(sdu_received, 0, 1);
+
+struct test_ctx {
+	struct bt_l2cap_le_chan le_chan;
+	size_t tx_left;
+} test_ctx;
+
+int l2cap_chan_send(struct bt_l2cap_chan *chan, uint8_t *data, size_t len)
+{
+	LOG_DBG("chan %p conn %u data %p len %d", chan, bt_conn_index(chan->conn), data, len);
+
+	struct net_buf *buf = net_buf_alloc(&sdu_pool, K_NO_WAIT);
+
+	if (buf == NULL) {
+		FAIL("No more memory\n");
+		return -ENOMEM;
+	}
+
+	net_buf_reserve(buf, BT_L2CAP_SDU_CHAN_SEND_RESERVE);
+	net_buf_add_mem(buf, data, len);
+
+	int ret = bt_l2cap_chan_send(chan, buf);
```

Comment:
I generally suggest to declare variables at top of scope (it's not an official coding guideline, so won't block the PR based on it)